crates/notary/server/src/tee.rs

```diff
@@ -0,0 +1,149 @@
+use mc_sgx_dcap_types::QlError;
+use serde::{Deserialize, Serialize};
+
+use crate::signing::AttestationKey;
+use p256::{ecdsa::SigningKey, PublicKey};
+use pkcs8::{DecodePrivateKey, EncodePrivateKey, LineEnding};
+use rand_chacha::{
+    rand_core::{OsRng, SeedableRng},
+    ChaCha20Rng,
+};
+use std::{
+    fs::{File, OpenOptions},
+    io::{self, Read, Write},
+    path::Path,
+};
+use tracing::{debug, error, instrument};
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Quote {
+    raw_quote: Option<String>,
+    mrsigner: Option<String>,
+    mrenclave: Option<String>,
+    error: Option<String>,
+}
+
+impl Default for Quote {
+    fn default() -> Quote {
+        Quote {
+            raw_quote: Some("".to_string()),
+            mrsigner: None,
+            mrenclave: None,
+            error: None,
+        }
+    }
+}
+
+impl std::fmt::Debug for QuoteError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            QuoteError::IoError(err) => write!(f, "IoError: {:?}", err),
+            QuoteError::IntelQuoteLibrary(err) => {
+                write!(f, "IntelQuoteLibrary: {}", err)
+            }
+        }
+    }
+}
+
+impl From<io::Error> for QuoteError {
+    fn from(err: io::Error) -> QuoteError {
+        QuoteError::IoError(err)
+    }
+}
+
+enum QuoteError {
+    IoError(io::Error),
+    IntelQuoteLibrary(QlError),
+}
+
+impl From<QlError> for QuoteError {
+    fn from(src: QlError) -> Self {
+        Self::IntelQuoteLibrary(src)
+    }
+}
+
+#[instrument(level = "debug", skip_all)]
+async fn gramine_quote() -> Result<Quote, QuoteError> {
+    //// Check if the the gramine pseudo-hardware exists
+    if !Path::new("/dev/attestation/quote").exists() {
+        error!("Failed to retrieve quote hardware");
+        return Err(QuoteError::IntelQuoteLibrary(QlError::InterfaceUnavailable));
+    }
+
+    // Reading attestation type
+    let mut attestation_file = File::open("/dev/attestation/attestation_type")?;
+    let mut attestation_type = String::new();
+    attestation_file.read_to_string(&mut attestation_type)?;
+    debug!("Detected attestation type: {}", attestation_type);
+
+    //// Writing 64 zero bytes to the gramine report pseudo-hardware `/dev/attestation/user_report_data`
+    let mut report_data_file = OpenOptions::new()
+        .write(true)
+        .open("/dev/attestation/user_report_data")?;
+    report_data_file.write_all(&[0u8; 64])?;
+
+    //// Reading from the gramine quote pseudo-hardware `/dev/attestation/quote`
+    let mut quote_file = File::open("/dev/attestation/quote")?;
+    let mut quote = Vec::new();
+    quote_file.read_to_end(&mut quote)?;
```

Comment:
question: is the quote measurement already available when this read happens? or this read ll trigger the measurement to happen?